sstables/compaction.cc

```diff
@@ -758,33 +762,40 @@ class regular_compaction : public compaction {
         return _creator();
     }
 
-    virtual sstable_writer* select_sstable_writer(const dht::decorated_key& dk) override {
-        if (!_writer) {
-            _sst = _creator();
-            setup_new_sstable(_sst);
+    virtual sstable_writer* select_sstable_writer(const dht::decorated_key& dk, compacting_sstable_writer *origin) override {
```

Comment:
Is this the correct place to make the decision? It is called once per partition key, but a TWCS reshaping may need to split partitions among multiple sstables.

Reply:
Note that the test I added explicitly deals with that, because it is the tricky scenario:
 it split a partition with two rows in multiple SSTables.

The reason that works is that the interposer "magically" generates two calls to consume_new_partition_key in that scenario.

That is why we are going through the interposer: the stream is known to be well behaved and we don't have to deal with corner cases in the writer.

Reply:
Although you are right that we can end up with many cells pertaining to different timestamps I think row granularity is good enough.

Reply:
> Although you are right that we can end up with many cells pertaining to different timestamps I think row granularity is good enough.

What do you mean "row granularity is good enough" - a row (CQL row?) may have different timestamps for different cells, tombstones, etc., which sstable do you pick - the first timestamp you see? Or what? I'm didn't understand what this code aims to do. I also didn't find the code which splits a partition into different rows going to different sstables, but it's proabably my fault for not understanding the code and I need to read it again.

Reply:
I mean a CQL row. What else is a row?

We have to make some compromise.
Time series workloads are usually of the form `(id, date)` as a key, so the partition key contains and ID and the row is pointed to by a date. That means that the chances of insertion happening for an entire row with the same timestamps is very high. So we split by CQL row.

You are right that it is technically possible that some application for whatever reason will have different cells with different timestamps, but as you noted yourself, the cost to process that is prohibitive.

In whichever case, the number of lines of code that would be changed in this patch if we wanted to process cells would be zero. And the reason for that is that I am reusing the interposer infrastructure

This patch just writes down whatever the interposer feeds us, and in doing that it layers things properly. I think that the interposer picks the same timestamp as the first cell - but to be honest I am not 100% sure, and I don't even really care (I do, just not for the perspective of this patch).

To see where this is split in multiple rows you should check the interposer code (start with sstables/compaction_strategy.cc, see the TWCS interposer overload). The interposer reads an incoming `flat_mutation_reader` stream - the one we generate from the SSTable set, and generates `n` output streams. All this patch does is wire up the interposer and write each stream individually.

Reply:
@nyh the code which actually splits mutations by timestamp is in https://github.com/scylladb/scylla/blob/master/mutation_writer/timestamp_based_splitting_writer.cc, the entry point is `segregate_by_timestamp()`. This is what the interposer @glommer talks about uses behind the curtain.

The resolution at which that works is cells, it examines every single timestamp in a mutation and routes the atom to the relevant consumer. If a row has 10 columns, each having a different timestamp, the splitter will create 10 rows with 1 columns each.

Reply:
@denesb good to know. However Nadav, note that even then, select_sstable_writer won't be called for every cell. The interposer will generate a different stream for each.

It does have to read the timestamps and act accordingly, but I suspect it will be cheaper than it would otherwise be on the compaction level.

We could do this just for major compaction but I'd like to see this done in a general. There is a cost associated with data inserted with the wrong timestamp - paid by readers - and it does happen. It doesn't happen often, but the users are unaware of it when it does happen and requiring a major compaction to fix it is bad. We're more resilient if we always compact correctly.

That said, this depends on the cost of the interposer, which as far as I understand nobody measured.

I will try to measure it for the next round.

Reply:
I don't understand this "select_sstable_writer". The splitter already creates writers as needed.

Reply:
Hello Avi

Maybe it can go away with further refactoring, but the reason for this is that right now this is chosen by the compaction object, of which there is only one.

The interposer code selects the writer based on `auto cr = c->get_compacting_sstable_writer();` so that the other strategies can keep working without much code change. The writers themselves are different objects (which is why I can use their addresses to differentiate the source of mutations), but they use infrastructure in the compaction object (like the current sstable object). I can try to refactor this by moving some of that inside the writer object. However my fear was that I'd have to end up writing an LCS interposer, which I fear will be more complex.

Note that `select_sstable_writer` is a pre-existing function, due to LCS.

Reply:
Resharding also uses it AFAIK.

Reply:
FWIW I just measured the cost of a TWCS compaction with and without the interposer.

The way I measured was to insert data into a TWCS table with a window period much larger than the insertion time. I then measured the total time to insert all the data + finish the remaining compactions.

The schema was a small-payload, 5 columns (the default cassandra-stress), which should generate a lot of per-metadata overhead and stress the CPU more than the I/O subsystem.

I didn't see a significant performance difference (< 1% total time)